hybrid/reference/pt.py

```diff
@@ -0,0 +1,223 @@
+            betas = [state.beta for state in states]
+
+        i = random.choice(range(len(states) - 1))
+        j = i + 1
+
+        return self.swap_pair(betas, states, i, j)
+
+
+class SwapReplicasDownsweep(SwapReplicaPairRandom):
+    """Parallel tempering swap replicas step.
+
+    On each call, sweep down and probabilistically swap all adjacent pairs
+    of replicas (input states).
+
+    Betas can be supplied in constructor, or otherwise they have to present in
+    the input states.
+
+    Args:
+        betas (list(float), optional):
+            List of betas (inverse temperature), one for each input state. If
+            not supplied, betas have to be present in the input states.
+    """
+
+    def __init__(self, betas=None, **runopts):
+        super(SwapReplicasDownsweep, self).__init__(betas=betas, **runopts)
+
+    def next(self, states, **runopts):
+        betas = self.betas
+        if betas is None:
+            betas = [state.beta for state in states]
+
+        for i in range(len(states) - 1):
+            states = self.swap_pair(betas, states, i, i + 1)
+
+        return states
+
+
+class SpawnParallelTemperingReplicas(hybrid.traits.SIMO, hybrid.Runnable):
+    """Expand the input state into N states enriched with betas. Betas are
+    calculated from the input problem.
+
+    Args:
+        num_replicas (int):
+            Number of replicas (output states) to create off of the input state.
+    """
+
+    def __init__(self, num_replicas, **runopts):
+        super(SpawnParallelTemperingReplicas, self).__init__(**runopts)
+        self.num_replicas = num_replicas
+
+    def next(self, state, **runopts):
+        bqm = state.problem
+
+        # get a reasonable beta range
+        beta_hot, beta_cold = neal.default_beta_range(bqm)
+
+        # generate betas for all branches/replicas
+        betas = np.geomspace(beta_hot, beta_cold, self.num_replicas)
+
+        # create num_replicas with betas spaced with geometric progression
+        states = hybrid.States(*[state.updated(beta=b) for b in betas])
+
+        return states
+
+
+#
+# A few PT workflow generators. Should be treated as Runnable classes
+#
+
+def ParallelTempering(num_sweeps=10000, num_replicas=10,
+                      max_iter=None, max_time=None, convergence=3):
+    """Parallel tempering workflow generator.
+
+    Args:
+        num_sweeps (int, optional):
+            Number of sweeps in the fixed temperature sampling.
+
+        num_replicas (int, optional):
+            Number of replicas (parallel states / workflow branches).
+
+        max_iter (int/None, optional):
+            Maximum number of iterations of the update/swaps loop.
+
+        max_time (int/None, optional):
+            Maximum wall clock runtime (in seconds) allowed in the update/swaps
+            loop.
+
+        convergence (int/None, optional):
+            Number of times best energy of the coldest replica has to repeat
+            before we terminate.
```

Comment:
"Repetitions of the best energy of the coldest replica that terminates the iterations"?